lib/src/actions/remote_backup_action.dart

```diff
@@ -3,60 +3,46 @@
 
 import 'dart:async';
 
-import '../transport/messages/backup_message.dart';
+import 'package:collection/collection.dart';
+import 'package:dart_2_party_ecdsa/dart_2_party_ecdsa.dart';
+import 'package:stream_transform/stream_transform.dart';
+
 import '../transport/shared_database.dart';
 
-class FetchRemoteBackupAction {
+class RemoteBackupListener {
   final SharedDatabase _sharedDatabase;
   final String _userId;
-  StreamSubscription<BackupMessage>? _streamSubscription;
-
-  final _completer = Completer<String>();
-
-  FetchRemoteBackupAction(this._sharedDatabase, this._userId);
-
-  Future<String> start() async {
-    _streamSubscription = _sharedDatabase
-        .backupUpdates(_userId)
-        .timeout(const Duration(seconds: 60))
-        .listen(_handleMessage, onError: _handleError, cancelOnError: true);
-    return _completer.future;
-  }
-
-  void cancel() {
-    _streamSubscription?.cancel();
+  final KeygenState _keygenState;
+  final BackupState _backupState;
+
+  RemoteBackupListener(this._sharedDatabase, this._userId, this._keygenState, this._backupState);
+
+  Stream<BackupMessage> start() {
+    return _sharedDatabase
+        .backupUpdates(_userId) //
+        .whereNotNull()
+        .tap(_handleMessage)
+        .handleError((error) {
+      print('Error listening remote backup: $error');
+    });
   }
 
   void _handleMessage(BackupMessage message) {
-    if (!_validateMessageDate(message)) {
-      return;
-    }
-    cancel();
-    _sharedDatabase.deleteBackupMessage(_userId);
-    _completer.complete(message.backupData);
-  }
-
-  void _handleError(Object error) {
-    _completer.completeError(error);
-  }
-
-  bool _validateMessageDate(BackupMessage message) {
-    final now = DateTime.now();
-    if (message.createdAt.add(message.expirationTimeout).isBefore(now)) {
-      return false;
-    } else {
-      return true;
+    if (message.address != null && message.walletId != null && message.backupData.isNotEmpty) {
```

Comment:
address and walletId will not be available in backup document for snap older then 1.2.9. We need to handle that case!

Reply:
walletId can be resolved